### DISCUSSION TITLE
How to draw color filled smooth closed spline curve through some points

### DISCUSSION THREAD

**Discussion by arnold408** [ImageMagick/ImageMagick#5259] [Help] [answered]
Hello,
Previously I had imagemagick 6 version and I used fmwconcepts spline script at
http://www.fmwconcepts.com/imagemagick/spline/index.php
Now I am using imagemagic 7 and I am not getting any output

I have tried this command to test the script
/root/spline -s kbs -e extend -d 141x101 "20,50 32.5,75 45,50 57.5,25 70,50 82.5,75 95,50 107.5,25 120,50"  /root/tests.jpg

There is no output error message and output image not showing

Please suggest how to fix or debug this.
Thanks

**Comment by fmw42** [comment]
Please supply your exact command line so that I can try to debug the issue.

**Reply by arnold408** [reply]
I have tried this command to test the script
/root/spline -s kbs -e extend -d 141x101 "20,50 32.5,75 45,50 57.5,25 70,50 82.5,75 95,50 107.5,25 120,50" /root/tests.jpg

There is no output error message and no output image

**Comment by snibgo** [comment]
@arnold408: please show the text output from "convert -version". We need to reproduce the problem, so please show your command line, and any input files.

From a quick look at Fred's spline script, I can't see any obvious problem with the fx expressions. But those depend on the values of $x0 etc.

**Reply by arnold408** [reply]
convert -version
Version: ImageMagick 7.1.0-35 Q16-HDRI x86_64 20062 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5)
Delegates (built-in): bzlib cairo djvu fontconfig freetype gslib gvc heic jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png ps raqm raw rsvg tiff webp wmf x xml zip zlib
Compiler: gcc (8.5)

**Comment by snibgo** [comment]
Update: it might be a double-minus problem, in lines like:
```
	t1x=`convert xc: -format "%[fx:($x1-$x0)]" info:`
```
... if $x0 is negative.

See one of my replies in the long fx thread: https://github.com/ImageMagick/ImageMagick/discussions/4533

For example:
```
magick xc: -format "%[fx:3--4]" info:

magick: Attempted update to non-UserSymbol '' at '4' @ error/fx.c/TranslateExpression/2570.
```
The problem is that "--" is interpreted as "subtract one from the previous variable". A fix is to use parentheses:

```
magick xc: -format "%[fx:3-(-4)]" info:

7
```

**Reply by arnold408** [reply]
Thanks for the reply. I have downloaded the script again and ran it. please see my question above. Thanks

**Comment by fmw42** [comment]
I can debug that if the OP posts his exact command line and original image.

**Reply by arnold408** [reply]
I have downloaded the script again, set 777 permission and ran the script again.
It gave no output image and no error message.
how to debug it?

I have tried other Fred's scripts and they are working fine without editing anything in the script.
Thanks

**Reply by arnold408** [reply]
I have tried this command to test the script
/root/spline -s kbs -e extend -d 141x101 "20,50 32.5,75 45,50 57.5,25 70,50 82.5,75 95,50 107.5,25 120,50" /root/tests.jpg

There is no output error message and no output image

**Reply by fmw42** [reply]
I think I have it fixed. Give me a few minutes to upload. I will let you know when to try to download again.

**Comment by fmw42** [comment, marked as answer]
try to download the script again from my web site, now. Then run the following or use a the list of control points as a string.

`spline -s kbs -t 0 -b 0 -e close -p skyblue -f im_logo_star.txt im_logo_star.gif im_logo_star_splinexx.gif`

![im_logo_star_splinexx](https://user-images.githubusercontent.com/13458218/175196132-f6789810-f329-4280-9d1e-fc5e89a341c7.gif)

**Comment by arnold408** [comment]
/root/spline -s kbs -e extend -d 141x101 "20,50 32.5,75 45,50 57.5,25 70,50 82.5,75 95,50 107.5,25 120,50"  /root/tests.jpg

Processing Segment 1 Out Of 8 Segments
Processing Segment 2 Out Of 8 Segments
Processing Segment 3 Out Of 8 Segments
Processing Segment 4 Out Of 8 Segments
Processing Segment 5 Out Of 8 Segments
Processing Segment 6 Out Of 8 Segments
Processing Segment 7 Out Of 8 Segments
Processing Segment 8 Out Of 8 Segments

convert: non-conforming drawing primitive definition `.3008' @ error/draw.c/RenderMVGContent/4490.

**Reply by fmw42** [reply]
Test the example on my web page. Does that work? Try integer values for your points.

**Reply by fmw42** [reply]
I am getting a different error and will have to research this.

**Comment by arnold408** [comment]
/root/spline -d 1537x1897 -s kbs -t 0 -b 0 -e close -bg white -i black "0,514 557,514 585,545 619,571 657,592 699,598 740,588 777,566 809,540 834,508 1437,508 1437,1797 0,1797" /root/test.jpg

Processing Segment 1 Out Of 13 Segments
Processing Segment 2 Out Of 13 Segments
Processing Segment 3 Out Of 13 Segments
Processing Segment 4 Out Of 13 Segments
Processing Segment 5 Out Of 13 Segments
Processing Segment 6 Out Of 13 Segments
Processing Segment 7 Out Of 13 Segments
Processing Segment 8 Out Of 13 Segments
Processing Segment 9 Out Of 13 Segments
Processing Segment 10 Out Of 13 Segments
Processing Segment 11 Out Of 13 Segments
Processing Segment 12 Out Of 13 Segments
Processing Segment 13 Out Of 13 Segments

realloc(): invalid next size
/root/spline: line 989: 4040101 Broken pipe             ( echo "viewbox 0 0 $width $height"; echo "image over 0,0 $width $height '$tmpA'"; if [ "$icolor" = "" ]; then
    echo "fill none stroke $ccolor stroke-width $strokewidth polyline $points";
else
    echo "fill $icolor stroke $ccolor stroke-width $strokewidth polyline $points";
fi; if [ "$spline" = "bezier" -a "$lcolor" != "" ]; then
    pArray=($plist); nps=${#pArray[*]}; num=`convert xc: -format "%[fx:2*$numseg]" info:`; nm1=`expr $num - 1`; i=0; j=1; c=0; skip=2; while [ $c -lt $num ]; do
        if [ $c -ne $nm1 ]; then
            lineArray[$c]="${pArray[$i]} ${pArray[$j]}!";
        else
            lineArray[$c]="${pArray[$i]} ${pArray[$j]}";
        fi; i=`expr $i + $skip`; j=`expr $j + $skip`; c=`expr $c + 1`; if [ $skip -eq 2 ]; then
            skip=1;
        else
            skip=2;
        fi;
    done; echo "fill none stroke $lcolor"; for theline in "${lineArray[*]}";
    do
        echo "$theline" | tr "!" "\012" | while read p1 p2; do
            echo "line $p1 $p2";
        done;
    done;
else
    if [ "$lcolor" != "" -a "$endmode" = "close" ]; then
        echo "fill none stroke $lcolor polygon $plist";
    else
        if [ "$lcolor" != "" -a "$endmode" != "close" ]; then
            echo "fill none stroke $lcolor polyline $plist";
        fi;
    fi;
fi; if [ "$pcolor" != "" ]; then
    echo "fill $pcolor stroke $pcolor"; echo "$plist" | tr " ," "\012 " | while read x y; do
        xx=`convert xc: -format '%[fx:'"$x"'+'"$prad"']' info:`; echo "circle $x,$y $xx,$y";
    done;
fi )
     4040102 Aborted                 (core dumped) | convert mvg:- "$outfile"

**Comment by fmw42** [comment]
I found a typo.  Try again.

`spline -s kbs -t 0 -p blue -e extend -d 141x101 "20,50 32.5,75 45,50 57.5,25 70,50 82.5,75 95,50 107.5,25 120,50" x.png`

![x](https://user-images.githubusercontent.com/13458218/175206890-f87cb22f-2278-4dd6-a3d2-0b5ea3afc396.png)


`spline -d 1437x1797 -s kbs -t 0 -b 0 -e close -bg white -i black "0,514 557,514 585,545 619,571 657,592 699,598 740,588 777,566 809,540 834,508 1437,508 1437,1797 0,1797" y.png`

![y](https://user-images.githubusercontent.com/13458218/175206899-b1ab1289-c833-4ed6-973c-54814d90320c.png)

**Comment by arnold408** [comment]
I have downloaded the script again and ran this command :
/root/spline -s kbs -t 0 -p blue -e extend -d 141x101 "20,50 32,75 45,50 57,25 70,50 82,75 95,50 107,25 120,50" /root/x2.png

Processing Segment 1 Out Of 8 Segments
Processing Segment 2 Out Of 8 Segments
Processing Segment 3 Out Of 8 Segments
Processing Segment 4 Out Of 8 Segments
Processing Segment 5 Out Of 8 Segments
Processing Segment 6 Out Of 8 Segments
Processing Segment 7 Out Of 8 Segments
Processing Segment 8 Out Of 8 Segments

convert: non-conforming drawing primitive definition `8' @ error/draw.c/RenderMVGContent/4490.

It gave blank image output

**Reply by fmw42** [reply]
Works perfectly fine for me.

`spline -s kbs -t 0 -p blue -e extend -d 141x101 "20,50 32,75 45,50 57,25 70,50 82,75 95,50 107,25 120,50" x.png`

![x](https://user-images.githubusercontent.com/13458218/175209205-e78c5bf3-a0b5-4e49-86db-ef58a88b522b.png)


Does it work for you if you do not run from root?  Perhaps in your home directory. Also perhaps preface with bash

bash spline -s kbs -t 0 -p blue -e extend -d 141x101 "20,50 32,75 45,50 57,25 70,50 82,75 95,50 107,25 120,50" x2.png

where you have spline script in some other directory and you cd to that directory to run it.

What is your IM version and platform/OS? What shell are you using?

**Reply by arnold408** [reply]
CentOS Stream Linux 8

convert -version
Version: ImageMagick 7.1.0-35 Q16-HDRI x86_64 20062 https://imagemagick.org/
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5)
Delegates (built-in): bzlib cairo djvu fontconfig freetype gslib gvc heic jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png ps raqm raw rsvg tiff webp wmf x xml zip zlib
Compiler: gcc (8.5)

**Comment by arnold408** [comment]
I have also tried outside root like this :

/home/wesbite.com/public_html/test/spline -s kbs -t 0 -p blue -e extend -d 141x101 "20,50 32,75 45,50 57,25 70,50 82,75 95,50 107,25 120,50" /home/wesbite.com/public_html/test/x2.jpg

got 
Processing Segment 1 Out Of 8 Segments
Processing Segment 2 Out Of 8 Segments
Processing Segment 3 Out Of 8 Segments
Processing Segment 4 Out Of 8 Segments
Processing Segment 5 Out Of 8 Segments
Processing Segment 6 Out Of 8 Segments
Processing Segment 7 Out Of 8 Segments
Processing Segment 8 Out Of 8 Segments

convert: non-conforming drawing primitive definition `3008' @ error/draw.c/RenderMVGContent/4490.

**Reply by fmw42** [reply]
I have no clue. I am on a Mac and it works fine. As I asked before can you put the script into another directory and cd to that directory and run without a path to the script name nor a path to your output? Can you preface with bash.

Is there more to the error message?  Does it give a line number?

What shell are you using in your terminal?  Is it bash or ksh or something else?

**Comment by arnold408** [comment]
uninstalled imageMagick 7
installed version 6.9
now it is working.
Thanks a lot

**Reply by fmw42** [reply]
Sorry, I did not know you were using IM 7.  I was testing in IM 6 and it works. Testing now under IM 7 with no error messages, but the result is not correct. I will track that down and see what is the issue.

**Reply by fmw42** [reply]
It is working for me with no error messages, but not drawing the lines. It does show the points.  So I have some work to track down the issue.

**Reply by fmw42** [reply]
Please in the future always identify your IM version and platform when asking questions.

**Reply by arnold408** [reply]
Thanks a lot for your great support.
I am extremely sorry for not mentioning IM version in the opening post.

**Reply by fmw42** [reply]
No problem.  Looks like the bug is in IM 7 end not my script. I will discuss with he IM developers.
